Add metric highlights

drewnoakes opened the pull request:
## Description

For the last month, @adamint and I have been working on `feature/dashpages`. We're now preparing it for review for merge into `main`.

A dashpage is a composite metric view, displayed in the metrics page when data for the relevant instruments are available.

Here's an example dashpage, showing multiple instruments in separate tiled charts:

![image](https://github.com/user-attachments/assets/06a6cf0f-e02d-4f7f-91d1-6b374b5039d6)

Currently dashpages are defined in a JSON file that ships with the dashboard. We include a default configuration.

Fixes #5289
Fixes #5295
Fixes #5303
Fixes #5298
Fixes #5287
Fixes #5347

## Checklist

- Is this feature complete?
  - [ ] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No, not initially.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5750)

drewnoakes committed:
Add "dashpage" as known dictionary word

drewnoakes committed:
Add "plotly" as a spelling word

drewnoakes committed:
Refactor markup

This will make it easier to add a block for rendering a dashpage.

drewnoakes committed:
Show dashpages in the metrics tree view

drewnoakes committed:
Add API doc

drewnoakes committed:
Add definition types for dashpage configuration

drewnoakes committed:
Add a collection of dashpages, and a selected dashpage

drewnoakes committed:
Allow a dashpage to be selected in the tree

drewnoakes committed:
Support rendering dashpages in content area

drewnoakes committed:
Add TODO and placeholder for DashpageChart component

drewnoakes committed:
Bump log messages to warnings

drewnoakes committed:
Make chart sizes responsive

adamint committed:
add DashpageCharts component

adamint committed:
add chartwrapper and tablewrapper components

drewnoakes committed:
Populate dashpages from configuration (#5302)

* Populate dashpages from configuration

Fixes #5289

This change causes the dashboard to look for a file named `dashpages.json` in the current directory. If found, it's parsed to provide the set of dashpages shown on the metrics page.

* Test fixes

bunit doesn't support when `OnInitializedAsync` actually yields the thread, so provide bunit tests with a stub implementation of `IDashpagePersistence.GetDashpagesAsync` that returns an empty set synchronously. In future we may extend this type to allow test setup with dashpage definitions.

drewnoakes committed:
Update default dashpages (#5320)

Fixes #5303

adamint committed:
isolate chartcontainer logic

adamint committed:
add ellipsis to dashpage instrument title

drewnoakes committed:
Filter dashpages to show only available ones (#5337)

The list of available dashpages in the Metrics UI is dynamic, depending on the selected resource.

A dashpage will only be shown if:

- metrics are available for at least one chart, and
- all required metrics are available
- all required resources are available

adamint committed:
Merge branch 'feature/dashpages' of https://github.com/dotnet/aspire into feature/dashpages

adamint committed:
show filter in popup

adamint committed:
Merge branch 'main' into feature/dashpages

# Conflicts:
#	src/Aspire.Dashboard/Components/Controls/Chart/ChartContainer.cs
#	src/Aspire.Dashboard/Components/Controls/Chart/PlotlyChart.razor
#	src/Aspire.Dashboard/Components/Controls/Chart/SingleMetricChartContainer.razor
#	src/Aspire.Dashboard/Components/Pages/Metrics.razor
#	src/Aspire.Dashboard/Components/Pages/Metrics.razor.cs
#	src/Aspire.Dashboard/Components/Pages/Metrics.razor.css
#	src/Aspire.Dashboard/Resources/Metrics.Designer.cs
#	src/Aspire.Dashboard/Resources/Metrics.resx
#	src/Aspire.Dashboard/wwwroot/js/app-metrics.js
#	src/Shared/StringComparers.cs

adamint committed:
fix merge issues

adamint committed:
dashpage ui work (#5693)

* fix merge, clean up

* save current dashpage

* beautify dashpages

* beautify chart filters in dashpages

* beautification

* fix bugs from merge

* improve table, chart appearance on mobile

* fix metric table height in dashpage, add dashpage header on mobile

* add fixedplacement to filter popover

* fix test

* continue to work on styling

* use unique ids for metric tables

* cleanup

* Fix broken tests

* Code formatting

* Avoid horizontal scrollbar in dashpage view

* Use integer for div id

This is consistent with other code that ensures unique IDs.

* Use correct string comparer

* Rename and doc property

---------

Co-authored-by: Drew Noakes <git@drewnoakes.com>

drewnoakes committed:
Merge remote-tracking branch 'upstream/main' into feature/dashpages

drewnoakes committed:
Formatting

drewnoakes committed:
Formatting

drewnoakes committed:
Remove unused component

drewnoakes committed:
Required parameter updates

drewnoakes committed:
Remove commented code

drewnoakes committed:
Remove unused variable

drewnoakes committed:
Use integer for unique element ID

drewnoakes committed:
Remove incorrect comment

drewnoakes committed:
Remove unneeded file

drewnoakes committed:
Simplify expression

drewnoakes committed:
[dashpages] Dashpage layout improvements (#5784)

* Use correct string comparer

* Make dashpage responsive to container, not viewport

`FluentGrid` uses `@media` queries for breakpoints, which work on the window size (viewport area). However we have a splitter in this UI, and dragging it should cause the dashpage area to relayout as the container size changes.

This change removes `FluentGrid` and instead uses flexbox layout directly.

maddymontaquila reviewed (commented):
Can I get a build of this or at least a demo vid before I give it a formal blessing???

leslierichardson95 commented:
> Can I get a build of this or at least a demo vid before I give it a formal blessing???

@adamint and @drewnoakes if possible, can either of you make it to a UX board review meeting and/or sync with David Culbertson this week?  I wanted to do one earlier but one of us was always OOF in the past few weeks. :)

JamesNK commented:
I haven't looked at any code yet. This is from testing.

Add a spacer between view dropdown and duration dropdown:
![image](https://github.com/user-attachments/assets/f913eaab-2d80-4f43-b601-a09b7e895d8d)
I think other toolbars use FluentDividers.

I expected the filters popup to be closable by clicking anyway on the page outside of the popup. This is how popups work everywhere else in Aspire (and web pages in general). However, only clicking in the panel will close the popup:
![dash-close-filter-popup](https://github.com/user-attachments/assets/8db2d1f5-c26a-4351-a87a-738108421944)

Multiple popups can be opened at the same time (would be fixed by fixing the issue above, because clicking outside of the popup would close it). Also, switching between views leaves the popup open (also would be fixed by the above):
![dash-popup-stays-open](https://github.com/user-attachments/assets/d7612b8f-9911-4330-966e-81f2cbdfa384)

Some other state is preserved when switching between views. "Show count" stays enable.

I don't know what happened, but while testing one of the charts stopped updating:
![dash-frozen](https://github.com/user-attachments/assets/d03aaeb5-aefb-4362-868e-f8c3f8d95777)
http.client.request.duration has frozen.

It might be related to this error in the console:
```
fail: Aspire.Hosting.Dashboard.Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost[111]
      Unhandled exception in circuit 'hYYAo8UMq3S58Ff2g215fTIFV_By0GPW3yIyuAYCwjo'.
      System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
         at System.Collections.Generic.List`1.Enumerator.MoveNext()
         at Aspire.Dashboard.Model.InstrumentViewModel.UpdateDataAsync(OtlpInstrumentSummary instrument, List`1 matchedDimensions) in C:\Development\Source\aspire\src\Aspire.Dashboard\Model\InstrumentViewModel.cs:line 23
         at Aspire.Dashboard.Components.ChartContainer.UpdateInstrumentDataAsync(OtlpInstrumentData instrument) in C:\Development\Source\aspire\src\Aspire.Dashboard\Components\Controls\Chart\ChartContainer.cs:line 113
         at Aspire.Dashboard.Components.ChartContainer.UpdateDataAsync() in C:\Development\Source\aspire\src\Aspire.Dashboard\Components\Controls\Chart\ChartContainer.cs:line 93
         at Aspire.Dashboard.Components.ChartContainer.DisposeAsync() in C:\Development\Source\aspire\src\Aspire.Dashboard\Components\Controls\Chart\ChartContainer.cs:line 70
         at Microsoft.AspNetCore.Components.RenderTree.Renderer.<>c__DisplayClass85_0.<<Dispose>g__HandleAsyncExceptions|1>d.MoveNext()
```

Opening the attribute popup inside the filter popup doesn't work:
![image](https://github.com/user-attachments/assets/43f8a4e1-3ef5-4348-85be-e850c6c3aad7)

I think there has been a regression on standalone metrics pages. The graph extends to fill the entire page. I believe it use to have a max width:
![image](https://github.com/user-attachments/assets/4a52f157-cb8e-4c14-84ef-1f1fbc913ec0)

I see the view title is in the query string. Is there some kind of internal identifer for views? It should be what is added to the query string. If there isn't an internal identifer for views then there probably should be one. For example, a view has a `Name` and a `DisplayName`. That way people can give views a unique identifer without worrying about conflicts, e.g. two people adding views titled "Database":
![image](https://github.com/user-attachments/assets/416afd56-6bb0-4f14-a075-b37f07e48740)

This button for a chart has filters and options. I don't think showing the filter button (highlighted in green) is right, specially since we're using it for another reason on the same screen. Since the popup is a mix of graph options, I think it should be a cog. A cog makes it clear that you're configuring the graph options.
![image](https://github.com/user-attachments/assets/42959740-1be7-4d5a-9fe3-6cc344106ed1)

The graph panels are hard up against the bottom of the page. The spacing between rows of panels should be added to the bottom:
![image](https://github.com/user-attachments/assets/173648f6-aefd-469f-991d-dcd3607df64f)

drewnoakes committed:
Merge branch main into feature/dashpages

adamint commented:
> How are dashpages sorted in the tree? Meters and instruments are alphabetical. Should dashpages be alphabetical?

We are displaying dashpages in the order that they are defined in `dashpages.json`. We could do either

adamint committed:
add header to dashpage desktop, add dashpage select page

adamint commented:
> An instrument page shows the instrument title and description. A dashpage doesn't, and you can only see what page you're on by looking at the tree. Should the dashpage title (and a description if there is one) be displayed? That would make it more consistent with other pages in the metrics area. And if we're going for a Grafana like feel, Grafana also displays the title at the top of a dashboard.

Previously, the title was just shown on mobile, now it's shown on desktop too. How do you like this?

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/97fa2719-a962-4e4d-a778-96890918f4b1">

Commit:
Invoke state after tree item selected

adamint commented:
> Clicking on dashpages menu option doesn't show anything, just the default "Select an instrument". That doesn't make sense in this situation. It should show a list of pages like clicking on a meter does for instruments:

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/2c925611-dbf4-4cc6-a65c-327555e0e204">

I reused our existing data grid for this!

adamint commented:
> The graph panels are hard up against the bottom of the page. The spacing between rows of panels should be added to the bottom:

Fixed. 
<img width="1285" alt="image" src="https://github.com/user-attachments/assets/66b0bf37-6f30-4fb9-af29-3b4522603d37">

> This button for a chart has filters and options. I don't think showing the filter button (highlighted in green) is right, specially since we're using it for another reason on the same screen. Since the popup is a mix of graph options, I think it should be a cog. A cog makes it clear that you're configuring the graph options.

Changed to a cog

Commit:
Add spacing below dashpage charts, change filter to cog

adamint commented:
> I see the view title is in the query string. Is there some kind of internal identifer for views? It should be what is added to the query string. If there isn't an internal identifer for views then there probably should be one. For example, a view has a Name and a DisplayName. That way people can give views a unique identifer without worrying about conflicts, e.g. two people adding views titled "Database":

Currently, there is no internal identifier for views. That display name is also unique

Commit:
Re-add area restricted to the dash chart container. This parameter is necessary for popovers within a card to display correctly. Sets fixedplacement in chart filter popup to allow breaking out of parent popup

adamint commented:
> http.client.request.duration has frozen. Closing and opening the filter popup causes values to update. I'm guessing an error was thrown and the timer that regularly updates data and renders stopped working. Closing and opening the filter causes a render and so data updates once.

I was unable to reproduce this issue. Do you have repro steps?

Commit:
Initialize ShowCounts in chart filter on component initialization - since this component will be initialized on each popup open

adamint commented:
> Add a spacer between view dropdown and duration dropdown:

Done. 
![image](https://github.com/user-attachments/assets/743c9e90-9785-4f82-8950-edf5009b1ae8)

> Some other state is preserved when switching between views. "Show count" stays enable.

I was able to reproduce this, now show count works properly inside the card.

Commit:
Add divider between chart type and time

Commit:
fix test by adding required member

Commit:
Fix accessibility regression on mobile

Commit:
Fix chart widths
- apply correct classes
- only use aspect-ratio and max-width on single metric view
- change the resize observer to watch the chart container itself. It was resizing to 100% because its *parent* takes up 100% of the width

adamint commented:
> I think there has been a regression on standalone metrics pages. The graph extends to fill the entire page. I believe it use to have a max width:

This has now been fixed. The new resize observer was resizing the plot to be larger than the maximum width we had set.
![image](https://github.com/user-attachments/assets/7bcc24fb-3cc3-4b2e-8b5c-2d85992a8cba)

JamesNK commented:
> Currently, there is no internal identifier for views. That display name is also unique

Please add one. Even if titles are unique now, this is going in the URL so there is the chance that people will bookmark it. It doesn't seem like a difficult change and it helps [avoid the URL changing in the future](https://www.w3.org/Provider/Style/URI).

JamesNK commented:
> I think this due to how cards work, and thus can't be changed without adding a document click listener. Is this a correct assumption @vnbaaij?

What are you using cards for that is special? If cards are a problem, why not use a regular div and apply the same styling?

JamesNK commented:
> I was unable to reproduce this issue. Do you have repro steps?

No. It looks like a thread safety problem: collection being modified at the same time as it is being enumerated

vnbaaij commented:
> > I think this due to how cards work, and thus can't be changed without adding a document click listener. Is this a correct assumption @vnbaaij?
> 
> What are you using cards for that is special? If cards are a problem, why not use a regular div and apply the same styling?

I generally don't use cards for a popup. For DataGrid resize popup we indeed just use a div with some script to detect clicking outside.

adamint commented:
> Multiple popups can be opened at the same time (would be fixed by fixing the issue above, because clicking outside of the popup would close it). Also, switching between views leaves the popup open (also would be fixed by the above):

This no longer seems possible

adamint commented:
> > Currently, there is no internal identifier for views. That display name is also unique
> 
> Please add one. Even if titles are unique now, this is going in the URL so there is the chance that people will bookmark it. Changing it to not use the title will break URLs.
> 
> It doesn't seem like a difficult change and it helps [avoid the URL changing in the future](https://www.w3.org/Provider/Style/URI).

Added

adamint committed:
Merge branch 'main' into feature/dashpages

# Conflicts:
#	src/Aspire.Dashboard/Components/Controls/Chart/ChartContainer.cs
#	src/Aspire.Dashboard/Components/Controls/Chart/ChartFilters.razor
#	src/Aspire.Dashboard/Components/Pages/Metrics.razor.cs

adamint committed:
fix conflict

Commit:
add import

radical commented:
/azp run

adamint committed:
additional dashpages changes (#6053)

* Separate dashpage id and display name

* sort dashpages consistently

* Change dashpages -> highlights

* Update dashpage query parameter to highlight, /dashpages -> /highlights

* update page directive

* change InstrumentViewModel.DataUpdateSubscriptions to prevent enumeration during collection modification, which can happen asynchronously

* Remove extra parameters after merge

* fix test

* update test expectation, remove comparer

* change dashpage -> highlight

* add warning comment

* put highlight id in path

---------

Co-authored-by: Adam Ratzman <adamratzman@microsoft.com>
